platform/ui/src/contextProviders/ViewportGridProvider.tsx

```diff
@@ -90,53 +101,61 @@ export function ViewportGridProvider({ children, service }) {
-        // and dropping a SEG/RT displaySet without any viewportOptions
-        // or displaySetOptions should not inherit the previous displaySet's
-        // which might have been a PDF Viewport. The viewport itself
-        // will deal with inheritance if required. Here is just a simple
-        // provider.
-        const viewport = state.viewports[viewportIndex] || {};
-        const viewportOptions = { ...payload.viewportOptions };
-
-        const displaySetOptions = payload.displaySetOptions || [];
-        if (displaySetOptions.length === 0) {
-          // Only copy index 0, as that is all that is currently supported by this
-          // method call.
-          displaySetOptions.push({ ...viewport.displaySetOptions?.[0] });
-        }
 
+      /**
+       * Sets the display sets for multiple viewports.
+       * This is a replacement for the older set display set for viewport (single)
+       * because the old one had a race conditions wherein the viewports could
+       * render partially in various ways causing exceptions.
+       */
+      case 'SET_DISPLAYSETS_FOR_VIEWPORTS': {
+        const { payload } = action;
         const viewports = state.viewports.slice();
 
-        let newView = {
-          ...viewport,
-          displaySetInstanceUIDs,
-          viewportOptions,
-          displaySetOptions,
-          viewportLabel: viewportLabels[viewportIndex],
-        };
-        viewportOptions.presentationIds = getPresentationIds(
-          newView,
-          viewports
-        );
-
-        // Make sure we assign a viewport id
-        newView = reuseViewport({}, newView, state.viewports);
-        console.log(
-          'Creating new viewport',
-          viewportIndex,
-          newView.viewportOptions.viewportId,
-          displaySetInstanceUIDs,
-          displaySetOptions
-        );
-
-        viewports[viewportIndex] = newView;
+        // Don't reuse any viewports here.
+        const idSet = state.viewports.reduce((accumulator, viewport) => {
+          accumulator[viewport.viewportOptions.viewportId] = true;
+          return accumulator;
+        }, {});
+
+        for (const updatedViewport of payload) {
+          // Note: there should be no inheritance happening at this level,
+          // we can't assume the new displaySet can inherit the previous
+          // displaySet's or viewportOptions at all. For instance, dragging
+          // and dropping a SEG/RT displaySet without any viewportOptions
+          // or displaySetOptions should not inherit the previous displaySet's
+          // which might have been a PDF Viewport. The viewport itself
+          // will deal with inheritance if required. Here is just a simple
+          // provider.
+          const { viewportIndex, displaySetInstanceUIDs } = updatedViewport;
+          const previousViewport = viewports[viewportIndex] || {};
+          const viewportOptions = { ...updatedViewport.viewportOptions };
```

Comment:
we say in the comment there should not be any inheritance, yet this is inheritance, either the comment is wrong or this line

Reply:
No, it is updatedViewport.viewportOptions, NOT previousViewport.viewportOptions, so it is simply updating the viewport.